lib/src/image.rs

```diff
@@ -2,33 +2,112 @@
+use clap::ValueEnum;
 use fn_error_context::context;
 use ostree_ext::container::{ImageReference, Transport};
+use serde::Serialize;
+use tabled::{settings::Style, Tabled};
 
-use crate::imgstorage::Storage;
+use crate::cli::{ImageListFormat, ImageListType};
 
 /// The name of the image we push to containers-storage if nothing is specified.
 const IMAGE_DEFAULT: &str = "localhost/bootc";
 
+#[derive(Clone, Tabled, Serialize, ValueEnum)]
+enum ImageListTypeColumn {
+    Host,
+    Logical,
+}
+
+impl std::fmt::Display for ImageListTypeColumn {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        self.to_possible_value().unwrap().get_name().fmt(f)
+    }
+}
+
+#[derive(Tabled, Serialize)]
+struct ImageOutput {
+    #[tabled(rename = "IMAGE TYPE")]
+    image_type: ImageListTypeColumn,
+
+    #[tabled(rename = "IMAGE")]
+    image: String,
+    // TODO: Add hash, size, etc? Difficult because [`ostree_ext::container::store::list_images`]
+    // only gives us the pullspec.
+}
+
+#[context("Listing host images")]
+fn list_host_images(sysroot: &crate::store::Storage) -> Result<Vec<ImageOutput>> {
+    let repo = sysroot.repo();
+    let images = ostree_ext::container::store::list_images(&repo).context("Querying images")?;
+
+    Ok(images
+        .iter()
+        .map(|x| ImageOutput {
+            image: x.to_string(),
+            image_type: ImageListTypeColumn::Host,
+        })
+        .collect())
+}
+
+#[context("Listing logical images")]
+fn list_logical_images(sysroot: &crate::store::Storage) -> Result<Vec<ImageOutput>> {
+    let output = sysroot
+        .get_ensure_imgstore()?
+        .new_image_cmd()?
+        .arg("list")
+        .arg("--format={{.Repository}}:{{.Tag}}")
+        .log_debug()
+        .output()?;
+
+    ensure!(
+        output.status.success(),
+        "Failed to list images: {}",
+        String::from_utf8_lossy(&output.stderr)
+    );
+
+    let stdout = String::from_utf8(output.stdout)?;
+
+    let images = stdout
+        .lines()
+        .map(|x| ImageOutput {
+            image: x.to_string(),
+            image_type: ImageListTypeColumn::Logical,
+        })
+        .collect();
+
+    Ok(images)
+}
+
 #[context("Listing images")]
-pub(crate) async fn list_entrypoint() -> Result<()> {
-    let sysroot = crate::cli::get_storage().await?;
-    let repo = &sysroot.repo();
+pub(crate) async fn list_entrypoint(
+    list_type: ImageListType,
+    list_format: ImageListFormat,
+) -> Result<()> {
+    // TODO: Get the storage from the container image, not the booted storage
+    let sysroot: crate::store::Storage = crate::cli::get_storage().await?;
 
-    let images = ostree_ext::container::store::list_images(repo).context("Querying images")?;
+    let images = match list_type {
+        ImageListType::All => {
+            let images = list_host_images(&sysroot)?;
+            let images = images.into_iter().chain(list_logical_images(&sysroot)?);
+            images.collect()
```

Comment:
This is fine as is but I would probably write this as
```
let mut images = list_host_images(&sysroot);
images.extend(list_logical_images(&sysroot)?);
images
```

Reply:
Went for

```rust
        ImageListType::All => list_host_images(&sysroot)?
            .into_iter()
            .chain(list_logical_images(&sysroot)?)
            .collect(),
```